File: 2020_2021/L3_S5_2020-2021/INFO0503_modélisation web serv & web avancé/INFO503/TP1-2/src/Voiture/Voiture.java
package Voiture;

import java.util.Date;

import Moteur.Moteur;

public class Voiture {
	private static int id = 0;
	private int idVoiture;
	private Modele modele;
	private Moteur moteur;
	private Date dateFab;
	private COULEUR couleur;
	private Option[] options;
	
	public Voiture(Modele mod, COULEUR couleur, Moteur mot, Option options[]) {
		this.idVoiture = id;
		Voiture.id ++;
		this.dateFab = new Date();
		this.modele = mod;
		this.couleur = couleur;
		this.moteur = mot;
		this.options = options;
	}
	
	public Voiture(Voiture v) {
		this.idVoiture = id;
		Voiture.id ++;
		this.dateFab = new Date();
		this.modele = v.getModele();
		this.moteur = new Moteur(v.getMoteur());
		this.couleur = v.getCouleur();
		this.options = new Option[v.getOptions().length];
		for(int i=0; i<v.options.length; i++) {
			this.options[i] = new Option(v.getOptions()[i].getNomOption(),v.getOptions()[i].getContainOption());
		}
	}
	
	public int getIdVoiture() {
		return this.idVoiture;
	}
	
	public Modele getModele() {
		return this.modele;
	}
	
	public Moteur getMoteur() {
		return this.moteur;
	}
	
	public Date getDateFab() {
		return this.dateFab;
	}
	
	public COULEUR getCouleur() {
		return this.couleur;
	}
	
	public Option[] getOptions() {
		return this.options;
	}
	
	public String getStringOptions() {
		String s = "";
		for(int i=0; i<this.options.length; i++) {
			s += this.options[i].getNomOption() + ", ";
		}
		return s;
	}
	
	public void setCouleur(COULEUR c) {
		this.couleur = c;
	}
	
	public String toString() {
		return ("Voiture : \n id : " + this.idVoiture + " ;\n Modele : " + this.modele.toString() + " ;\n Couleur : " + this.couleur 
				+ " ;\n Date Fabrication : " + this.dateFab.toString() + " ;\n Moteur : " + this.moteur.toString() 
				+ " ;\n Options : " + this.getStringOptions());
	}
}

File: 2020_2021/L3_S5_2020-2021/INFO0503_modélisation web serv & web avancé/INFO503/TP2/Voiture/Voiture.php
<?php
include("/UwAmp/www/INFO503/TP2/Moteur/Moteur.php");
include("/UwAmp/www/INFO503/TP2/Voiture/Options.php");
include("/UwAmp/www/INFO503/TP2/Voiture/Modele.php");
include("/UwAmp/www/INFO503/TP2/Voiture/Couleur.php");

class Voiture implements JsonSerializable{
    protected $modele;
    protected $moteur;
    protected $dateFabrication;
    protected $couleur;
    protected $options;
    protected $idVoiture = -1;
    static $id = 0;
    
    public function __construct($modele, $moteur, $couleur, $options, $dateFabrication, $idVoiture = 0){
        if($idVoiture != 0){$this->idVoiture = $idVoiture;}
        else {
            self::$id++;
            $this->idVoiture = self::$id;
        }
        $this->modele = $modele;
        $this->moteur = $moteur;
        $this->dateFabrication = $dateFabrication;
        $this->couleur = $couleur;
        $this->options = $options;
    }

    public function getIDVoiture() {return $this->idVoiture;}
    public function getMoteur() {return $this->moteur;}
    public function getModele() {return $this->modele;}
    public function getDateFabrication() : DateTime {return $this->dateFabrication;}
    public function getCouleur() : String {return $this->couleur;}
    public function getOptions() : Options {return $this->options;}

    public function setCouleur(String $couleur = Couleur::BEIGE){$this->couleur = $couleur;}
    public function setOption($options){$this->options = $options;}

//Méthodes
    public function __toString() : String {
        $thisModele = $this->getModele();
        $thisMoteur = $this->getMoteur();
        return "id : " . $this->getIDVoiture() 
        //. $this->getModele()->prout()
        . " modele : " . $thisModele->__toString()
        . " moteur : " . $thisMoteur->__toString() 
        . " date de fabrication : " . date_format($this->getDateFabrication(), 'Y-m-d') 
        . " couleur : " . $this->getCouleur() 
        . " options : " . $this->getOptions()->__toString();
    }
    
    public function jsonSerialize() : array {
        return ['idVoiture' => $this->getIDVoiture(), 'modele'=> $this->getModele(), 'moteur' => $this->getMoteur(), 'dateFabrication' => $this->getDateFabrication(), 'couleur' => $this->getCouleur(), 'options'=> $this->getOptions()];
    }

    public static function FromJson($json) : Voiture {
        $table = json_decode($json);
        return new self(
            new Modele($table->modele->nom)
            , new Moteur($table->moteur->carburation, $table->moteur->puissance, $table->moteur->idMoteur)
            , $table->couleur
            , new Options($table->options->GPS, $table->options->VitreElectriques, $table->options->Climatisation, $table->options->CameraDeRecul, $table->options->ToitOuvrant, $table->options->Alarme, $table->options->AirBagLateraux)
            , new DateTime($table->dateFabrication->date, new DateTimeZone($table->dateFabrication->timezone))
            , $table->idVoiture);
    }
}
?>